kernel/kbd.c

```diff
@@ -54,23 +54,36 @@ void read_key (void) {
     /* Don't */
     r = scancode[c];
   }
-  /* Put the character in the buffer */
-  buf[buf_head] = r;
-  buf_head = (buf_head + 1) & 255;
-  if (CIRC_SPACE(buf_head, buf_tail, 256) == 0) {
-    /* Buffer's full, so drop least-recent char. */
-    buf_tail = (buf_tail + 1) & 255;
+
+  tcb *t = (tcb *)list_pop_front(&kbd_wait);
+  if (t != 0) {
+    /* Threads are waiting; give the first one in line a character. */
+    if (CIRC_CNT(buf_head, buf_tail, 256) > 0) {
+      panic("Logic error in kbd.c; both threads and characters waiting.");
+    }
+    list_remove(&t->wait_queue);
+    t->saved_registers->rax = r;
+    reschedule_thread(t);
+  } else {
+    /* No threads waiting, buffer the character. */
+    buf[buf_head] = r;
+    buf_head = (buf_head + 1) & 255;
```

Comment:
% 256?

Reply:
Done, and also in the CIRC_CNT macro definition above.  (Which means that CIRC_CNT now works for non-power-of-two sizes)